python/lsst/daf/butler/direct_query_driver/_driver.py

```diff
@@ -0,0 +1,1200 @@
+                partition_by=partition_by, order_by=rank_sql_column
+            )
+        else:
+            builder.joiner.special["_ROWNUM"] = sqlalchemy.sql.func.row_number().over(
+                order_by=rank_sql_column
+            )
+        # ... and then turn that into a subquery with a constraint on rownum.
+        builder = builder.nested(force=True)
+        # We can now add the WHERE constraint on rownum into the outer query.
+        builder.joiner.where(builder.joiner.special["_ROWNUM"] == 1)
+        # Don't propagate _ROWNUM into downstream queries.
+        del builder.joiner.special["_ROWNUM"]
+        return builder
+
+    def _analyze_query_tree(self, tree: qt.QueryTree) -> tuple[QueryJoinsPlan, QueryBuilder]:
+        """Start constructing a plan for building a query from a
+        `.queries.tree.QueryTree`.
+
+        Parameters
+        ----------
+        tree : `.queries.tree.QueryTree`
+            Description of the joins and row filters in the query.
+
+        Returns
+        -------
+        plan : `QueryJoinsPlan`
+            Initial component of the plan relevant for the "joins" stage,
+            including all joins and columns needed by ``tree``.  Additional
+            columns will be added to this plan later.
+        builder : `QueryBuilder`
+            Builder object initialized with overlap joins and constraints
+            potentially included, with the remainder still present in
+            `QueryJoinPlans.predicate`.
+        """
+        # Delegate to the dimensions manager to rewrite the predicate and start
+        # a QueryBuilder to cover any spatial overlap joins or constraints.
+        # We'll return that QueryBuilder at the end.
+        (
+            predicate,
+            builder,
+        ) = self.managers.dimensions.process_query_overlaps(
+            tree.dimensions,
+            tree.predicate,
+            tree.get_joined_dimension_groups(),
+        )
+        result = QueryJoinsPlan(predicate=predicate, columns=builder.columns)
+        # Add columns required by postprocessing.
+        builder.postprocessing.gather_columns_required(result.columns)
+        # We also check that the predicate doesn't reference any dimensions
+        # without constraining their governor dimensions, since that's a
+        # particularly easy mistake to make and it's almost never intentional.
+        # We also allow the registry data ID values to provide governor values.
+        where_columns = qt.ColumnSet(self.universe.empty.as_group())
+        result.predicate.gather_required_columns(where_columns)
+        for governor in where_columns.dimensions.governors:
+            if governor not in result.constraint_data_id:
+                if governor in self._defaults.dataId.dimensions:
+                    result.constraint_data_id[governor] = self._defaults.dataId[governor]
+                else:
+                    raise qt.InvalidQueryError(
```

Comment:
It's fine for the same error to be raised either from client-side and server-side, and to me this makes sense as an `InvalidQueryError`.

The main distinction we need to make is between:

"this is a thing that happened because of the arguments the user specified, and the only way to fix it is for them to change their query, or add some more data to the server, or..." -> subclass of ButlerUserError, so it's re-raised on the client side.

vs

"this is a problem with the database server, an internal logic problem, an unexpected assertion failure, etc". -> not a subclass of ButlerUserError, so that the details aren't sent to the client and it triggers alerts for our team to investigate.